### DISCUSSION TITLE
Ecosystem ESM migration in the context of TypeScript

### DISCUSSION THREAD

**Discussion by jgoux** [rollup/plugins#1198] [General]
Hello all 👋 

I'm working on a monorepo using TypeScript and ESM. I'm having trouble with many external packages that export both ESM/CJS formats, but often the types generated for the ESM part aren't correctly generated. In the majority of the cases, they lack the full path filename in the import/export statement, for example: `import { foo } from './foo'` doesn't work, it should be `import { foo } from './foo.js'`, without the full path, TypeScript doesn't see the import in the exposed `.d.ts` declaration files.

Right now I'm opening issues and PR and suggest adding the extensions explicitly. I was wondering if there is an easier way through a rollup plugin maybe (the majority of the packages are using rollup for building the project)? Is the typescript plugin able to add the extensions (it should also handle adding the `index.js`) when compiling a project? Maybe it's just an option they don't know and the fix would be painless! 😄 

Example of projects exporting broken type definitions in ESM and using rollup:
- https://github.com/crutchcorn/cli-testing-library
- https://github.com/tinyhttp/tinyhttp

Thanks for your help!

**Edit**:

I was wrong, no need for anything else, please see: https://github.com/rollup/plugins/discussions/1198#discussioncomment-2856136

**Comment by crutchcorn** [comment]
This may potentially be impacting of a significantly larger project than either my library or TinyHTTP as well. I based my library's codebase structure off of [`dom-testing-library`](https://github.com/testing-library/dom-testing-library), a popular testing library.

@jgoux have you had the chance to test against that library to see if it has the same issues as CLI Testing Library?

**Reply by jgoux** [reply]
Indeed, I just tested and `dom-testing-library` has the same issue.

**Comment by jgoux** [comment]
I'm a bit confused because I'm also using `@faker-js/faker` and they don't specify explicit extensions, but I have no issue consuming their package. 🤔 https://unpkg.com/browse/@faker-js/faker@7.1.0/dist/types/index.d.ts

I'm still trying to figure out why it is working for them but not for you @crutchcorn

**Reply by jgoux** [reply]
In fact I was wrong all along. 🥇

There is no issue consuming CSJ packages exporting their types definitions without extensions.

In the case of https://github.com/crutchcorn/cli-testing-library it was an incorrect type definition that makes me think the issue was on the imports. (and I said `@testing-library/dom` had the same issue because they didn't use extensions as well, but in fact it's working just fine as they're a CSJ package!)

In the case of https://github.com/tinyhttp/tinyhttp which is a pure ESM package, the extensions are required (and my PR is up!) 👍
